### ISSUE TITLE
[1.9] Random: 'MongoTimeoutError: Server selection timed out after 10000 ms'

### ISSUE THREAD

**Issue by KoenLav** [meteor/meteor#10924]
Some random amount of time after deploying our application to staging with Meteor 1.9 (connected to a MongoDB 4.2.3 Sandbox instance on Atlas) the server will start producing the following error:

`MongoTimeoutError: Server selection timed out after 10000 ms`

I believe we might be experiencing this issue:
https://github.com/Automattic/mongoose/issues/8381

I realize Meteor 1.9 is not supposed to support MongoDB 4.2.3 yet, so I'm deploying with 1.10 beta and/or useUnifiedTopology: false now to see if that resolves things, but I figured others may be running into the same issue, hence I opened this one for reference.

**Comment by KoenLav** [comment]
Deploying with useUnifiedTopology set to false seems to have resolved the issue.

Not sure what other things may break. I'll try deploying with Meteor 1.10 tomorrow.

**Comment by KoenLav** [comment]
Possibly related to:

https://jira.mongodb.org/browse/NODE-2274

**Comment by KoenLav** [comment]
Deploying with Meteor 1.10 also seems to resolve the issue (which makes sense, since the bugfix for the bug mentioned above is included in the mongodb driver version included in Meteor 1.10).

What is strange though: we are not experiencing this issue on another application which has also been upgraded to Meteor 1.9 and is connecting to the exact same database...

**Comment by ramzauchenna** [comment]
Started having this issue after deploying to latest node and latest meteor patch

**Comment by filipenevola** [comment]
@ramzauchenna could you try meteor update --release 1.10-rc.1?

**Comment by doronrk** [comment]
Not here just to say "me too", but also to urge a fix in 1.9.x series since this is causing problems for us in production and I'd rather not tack on a substantial upgrade to fix this problem.

**Comment by doronrk** [comment]
@KoenLav Looking at release-1.9.2 it looks like `useUnifiedTopology` isn't actually set anywhere in meteor. The mongo documentation is a little vague, but it seems like `useUnifiedTopology` defaults to false? 

Can you check my understanding here? Trying to figure out why setting `useUnifiedTopology` would change any behavior on 1.9.2

**Comment by jnurkka** [comment]
> Not here just to say "me too", but also to urge a fix in 1.9.x series since this is causing problems for us in production and I'd rather not tack on a substantial upgrade to fix this problem.

I totally agree. Also, given the issues upgrading to 1.9. caused us due to this, I'd like to test 1.10. in a staging environment for at least a month. And not keen to have to work with 1.8 for another month

> @KoenLav Looking at release-1.9.2 it looks like `useUnifiedTopology` isn't actually set anywhere in meteor. The mongo documentation is a little vague, but it seems like `useUnifiedTopology` defaults to false?

I'd add that our staging environment just experienced this same issue even though we've set `useUnifiedTopology` to `false`

**Comment by doronrk** [comment]
@jnurkka Thanks for this valuable info.

> I'd add that our staging environment just experienced this same issue even though we've set useUnifiedTopology to false

Yeah this makes complete sense. I've confirmed that `useUnifiedTopology` defaults to false looking at the 3.2.7 driver code.

**Comment by KoenLav** [comment]
@doronrk when I search through the Meteor repository I can't find it either, but when I search through my .meteor/local directory in a project running Meteor 1.9 I do find this:
```
      // Required to silence deprecation warnings with mongodb@3.2.1.
      useUnifiedTopology: true
```

**Comment by generalleger** [comment]
@KoenLav That is very helpful. Just to be clear, wrapping the following in a Meteor.startup function should do the trick?

`Mongo.setConnectionOptions({ useUnifiedTopology: false });`

**Comment by KoenLav** [comment]
I put it before (outside of) Meteor.startup on our server code.

I guess the config should be set before the database connection is initialized.

**Comment by generalleger** [comment]
Ahh okay, i was thinking to do it after so that the default initialization process would not override the database connection setting!  To be safe, I'll do both until a formal fix comes out.

**Comment by jnurkka** [comment]
> I put it before (outside of) Meteor.startup on our server code.
> 
> I guess the config should be set before the database connection is initialized.

I also put it as early as possible in the server code, outside the Meteor.startup callback. This didn't seem to work as I stated above.

**Comment by doronrk** [comment]
@KoenLav thanks for sharing your finding.

> but when I search through my .meteor/local directory in a project running Meteor 1.9 I do find this:       
// Required to silence deprecation warnings with mongodb@3.2.1.
      useUnifiedTopology: true

The .meteor/local directory contains a bunch of stuff, including old builds which may have been built with different versions (in this case, we can be certain this is what is happening, because the comment string you posted appears in `1.10-rc.3` and is from [this commit](https://github.com/meteor/meteor/commit/282b16ed16f416d0e399657348616cc3ff408ea8#diff-a9d0b23557a439d95f0ceea868fb961b)

The source of truth to answer the question "does meteor set useUnifiedTopology to true in 1.9 or 1.9.2" is the source code on `release-1.9` and `release-1.9.2`. You can see [here](https://github.com/meteor/meteor/blob/release-1.9.2/packages/mongo/mongo_driver.js) that `useUnifiedTopology` is not set. You can also clone the meteor repo, checkout `release-1.9.2` and grep for `useUnifiedTopology: true` and you'll only find it in the actual `mongodb` and `mongodb-core` packages.

@jnurkka 's report that this did not solve his problem in practice, combined with the above to me suggests that there is still not a reliable fix for this issue.

**Comment by doronrk** [comment]
For posterity. 1.9.3-rc.0 is being released with updated mongo drivers

**Comment by KoenLav** [comment]
So maybe I was unclear, but you should set useUnifiedTopology to false, not true. 

I'm fairly certain this resolved the issue we were experiencing. But either way, as long as this gets resolved somehow, I'm fine :)

**Comment by ramzauchenna** [comment]
@filipenevola the issue stopped after i reverted to node 12.14.1. No idea if this fixed the issue but have not experienced it after this and i am using meteor 1.9.2

**Comment by filipenevola** [comment]
Hi @doronrk any news on 1.9.3-rc.0? Is the new driver fixing your issue?

The PR is here https://github.com/meteor/meteor/pull/10961

**Comment by cfnelson** [comment]
Can confirm for anyone else that updating from Meteor 1.9.2 to -> 1.9.3.rc.0 has fixed the issue for us (hosting on Meteor Galaxy).
Updating to the 1.9.3 release were the only changes made to resolve the problem.

For anyone in the future -> `meteor update --release 1.9.3-rc.0` or once v1.9.3 lands, just run `meteor update`  and redeploy your app.

**Comment by filipenevola** [comment]
Fixed on 1.9.3
